Fix verifyDelegator() logic in the allow handler

dmitrizagidulin opened the pull request:
Fixes issue #473.

melvincarvalho commented:
I can confirm that sending the on-behalf-of header no longer crashes the server

dmitrizagidulin commented:
@melvincarvalho Thanks for confirming!

dmitrizagidulin commented:
@dan-f / @nicola / @deiu  - ready for review.

csarven commented:
Confirmed.

dan-f commented on lib/handlers/allow.js:
```diff
@@ -12,59 +15,65 @@ process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
 
 function allow (mode) {
```
Could you add a `JSDoc` comment for this function?

dan-f commented on lib/handlers/allow.js:
```diff
@@ -93,56 +102,69 @@ function fetchDocument (host, ldp, baseUri) {
-      if (utils.debrack(delegateeValue.toString()) === delegatee) {
-        callback(null, true)
+/**
+ * @see https://www.w3.org/wiki/WebID/Delegation
```
Awesome `JSDoc` - the `@see` block tag is new to me. I think this could still use a brief one-liner to the top of the comment to introduce the function.

dan-f commented on lib/handlers/allow.js:
```diff
@@ -93,56 +102,69 @@ function fetchDocument (host, ldp, baseUri) {
+          ns.acl('delegates'),
+          rdf.namedNode(secretary)
+        )
+      let isAuthorized = authorizedMatches.length > 0
```
Great. I think you could squish these two statements into one using the `.any()`.

dan-f commented on lib/handlers/allow.js:
```diff
@@ -93,56 +102,69 @@ function fetchDocument (host, ldp, baseUri) {
+ * @return {Promise<string>} The user id (either the current user, or
+ *   the user being delegated for).
+ */
+function getUserId (webId, onBehalfOf, fetchDocument) {
```
I think this function could use a slightly more specific name. Something like `getRequestingWebId`?

Another thing is that I think the `webId` and `onBehalfOf` parameters are an implementation concern of this function rather than the caller. Could we just pass the request object instead?

dan-f commented on lib/handlers/allow.js:
```diff
@@ -93,56 +102,69 @@ function fetchDocument (host, ldp, baseUri) {
+ */
+function getUserId (webId, onBehalfOf, fetchDocument) {
+  let principal = onBehalfOf
+  if (!webId || !principal) {
```
I don't understand this logic. If there's no webId in the request, resolve to the webId? That seems like an error condition to me.

dan-f replied:
Or does that to handle unauthenticated users? If we're trying to return a null value for when the request is unauthenticated I'd prefer to explicitly return `null`.

dmitrizagidulin replied:
Yeah, this is meant to handle un-authenticated users. Sure, we can return `null` explicitly, no prob.

dan-f commented on lib/utils.js:
```diff
@@ -19,8 +19,15 @@ var S = require('string')
+ */
 function debrack (s) {
-  if (s.length < 2) {
+  if (!s || s.length < 2) {
```
Not to be annoying, but this should have a test case :)

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+describe('allow() handler', () => {
+  describe('getUserId()', () => {
+    const emptyFetchDocument = () => {}
+    it('no webId present', done => {
```
I have a strong preference for test cases to be named after the behavior that they're looking for rather than the data they accept. For example, rather than "no webId present" something like "should <do expected behavior> when no webId is present". This clarifies the intended logic for contributors, future developers, and us months down the line when we forget what the intended behavior was.

dan-f commented on lib/handlers/allow.js:
```diff
@@ -12,59 +15,65 @@ process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
+        })
+      })
+      .catch(err => {
+        next(err)
```
I'm not too familiar with express. Does this pass the `err` object to the next middleware?

dmitrizagidulin replied:
Yeah, `next()` is Express's generic error and flow control handler. If not handled by the next middleware layer, it ends up being displayed to the user in `error-pages.js` (which is usually the desired behavior).

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+const notVerifiedSecretaryGraph = rdf.graph()
+
+describe('allow() handler', () => {
+  describe('getUserId()', () => {
```
I expected there to be two separate (not nested) describe blocks - one for the `allow` function and one for the `getUserId()` function - since they are separate exported functions from the `allow` module.

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+    const emptyFetchDocument = () => {}
+    it('no webId present', done => {
+      let webId  // no secretary
+      let onBehalfOf = alice  // principal
```
Additionally, I expect the case in which the request has no webId but an on-behalf-of header to be a rejected promise, not for `getUserId` to resolve to `undefined`.

As an aside I'd generalize that I'd almost never expect a promise to resolve to `undefined`. Null or falsy values should be explicit.

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+        .then(userId => {
+          assert.equal(userId, undefined,
+            'Should return undefined when no webId is present')
+          done()
```
Rather than accept the `done` callback and imperatively call it, mocha lets you return a promise from your test function.

dmitrizagidulin replied:
I've heard that as well. But in testing, returning the promise and not calling `done()` only results in `Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.` :(

dan-f replied:
You have to 1) remove the `done` parameter from the test function and 2) return the promise from the test function.

dmitrizagidulin replied:
Ah I see, yeah, I was leaving in the `done` parameter to the test.

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+          done()
+        })
+    })
+    it('no onBehalfOf present', done => {
```
Same comment regarding test case naming.

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+          done()
+        })
+    })
+    it('throws 500 error if fetchDocument errors', done => {
```
I think its an anti-pattern to knowingly allow and test for 500 errors in a web server. If there's a case in which you know you cannot complete a request you've got a slew of semantic error codes to choose from.

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+        .catch(err => {
+          console.error(err)
+        })
+    })
```
There needs to be a test case for an entirely unauthenticated request - no webId and no on-behalf-of header.

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+            'Should return principal (onBehalfOf) if secretary is verified')
+          done()
+        })
+        .catch(err => {
```
You should remove this `catch` block which will suppress errors from making it back to mocha in your test.

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+            'Should return the webId (secretary id) if secretary is NOT verified')
+          done()
+        })
+        .catch(err => {
```
Remove this catch block.

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+  })
+
+  describe('verifyDelegator()', () => {
+    it('should throw 500 error if fetchDocument errors', done => {
```
Again, can we come up with a more semantic error code?

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+            'Should be true if profile authorizes the secretary')
+          done()
+        })
+        .catch(err => {
```
remove this `catch` block

dan-f commented on test/allow.js:
```diff
@@ -0,0 +1,131 @@
+            'Should be false if profile does not authorize the secretary')
+          done()
+        })
+        .catch(err => {
```
remove this `catch` block

dmitrizagidulin commented:
> The documentFetcher function isn't tested. Since it's basicly just a web client, rather than write a bunch of tests for it, couldn't we use the solid web client directly?

As much as I would *love* to refactor `documentFetcher()` to use solid-web-client (or something similarly reusable), that's a fairly involved task and out of scope for this fix/PR. It is a pain point for me, though, and I will refactor it soon (probably when integrating the ad hoc Groups permission support).

dmitrizagidulin commented:
PR feedback addressed in the latest commit.

- Added docstring to `allow(mode)` handler.
- Renamed `getUserId()` to `getRequestingWebId()`
- Returning `null` explicitly in `getRequestingWebId()`.
- Updated the test description strings
- Removed explicit testing for 500 errors
- Removed `.catch()` blocks in the tests.
- Added tests for utils/debrack()

> The `allow` handler itself isn't tested but should be. It would be a good use case for a http mocking library such as nock. Perhaps this is more of an integration level test?

It's being tested on an integration leve in `test/acl.js`.

> Are we running standard on the tests? I'd argue we should.

Yes, our `npm test` also invokes `standard` on the tests directory.

> The lack of newlines between test cases makes them a bit hard to read for me.

No prob, added more newlines.

> The `documentFetcher` function isn't tested. Since it's basicly just a web client, rather than write a bunch of tests for it, couldn't we use the solid web client directly?

Yeah, refactoring the `documentFetcher()` function is on my to-do list (it's used extensively in `acl-checker.js`, and tested indirectly in `test/acl.js`). Writing tests for it is out of scope for this PR.

nicola commented on lib/handlers/allow.js:
```diff
@@ -10,61 +13,84 @@ var utils = require('../utils')
+          : req.path
+        ldp.exists(req.hostname, reqPath, (err, ret) => {
+          if (ret) {
+            var stat = ret.stream
```
I know it was there before, but, where is this used?

dmitrizagidulin replied:
@nicola No idea.

dan-f commented on test/allow.js:
```diff
@@ -14,117 +14,94 @@ verifiedSecretaryGraph.add(
-          done()
-        })
-    })
-    it('throws 500 error if fetchDocument errors', done => {
```
What happened to this test?

dmitrizagidulin replied:
I removed that test because the fetch document function callback is no longer throwing a 500. I realized that the callback had no way of knowing the semantics of why it failed, whether for a 4xx-type reason or a 5xx-type. 
It'll be the responsibility of `fetchDocument()`, when we refactor it, to provide a more semantic error code.

dan-f replied:
As noted below, I think that if the logic in this PR depends on an untested function, it is necessarily in scope for that function to be tested.

dmitrizagidulin replied:
*Many* functions inside solid-server, on which this PR depends, are untested. Blocking a fix until the developer adds tests to all of them is unrealistic.

dan-f commented on test/allow.js:
```diff
@@ -14,117 +14,94 @@ verifiedSecretaryGraph.add(
   })
 
   describe('verifyDelegator()', () => {
-    it('should throw 500 error if fetchDocument errors', done => {
```
Why remove these tests? My comment was about testing for a more semantic error status code, not removing the tests.

dan-f commented on test/utils.js:
```diff
@@ -31,4 +31,27 @@ describe('Utility functions', function () {
+  describe('debrack()', () => {
+    it('should return null if no string is passed', done => {
+      assert.equal(utils.debrack(), null)
+      done()
```
You don't need to accept/call the `done` parameter in synchronous tests.

dmitrizagidulin replied:
Thanks. Removed.

dan-f commented on lib/handlers/allow.js:
```diff
@@ -108,20 +125,31 @@ function documentFetcher (host, ldp, baseUri) {
-function getUserId (webId, onBehalfOf, fetchDocument) {
-  let principal = onBehalfOf
-  if (!webId || !principal) {
+function getRequestingWebId (webId, onBehalfOf, fetchDocument) {
```
My comment still stands that I think this function would be cleaner if it accepted the request object rather than `webId` and `onBehalfOf`.

dmitrizagidulin replied:
I disagree. That function has no need of the rest of the `req` object. And in fact, does not care which headers the Web ID or the secretary id live.

dan-f replied:
Some procedure has to know where that data comes from, and rather than have every single caller of this function reach into the request object and pull out those values, it seems more maintainable to keep that logic here.

dmitrizagidulin replied:
Again, I disagree. And there's only one caller of this function.

dan-f commented:
Looks like my last few comments got squished as "outdated". Not sure why, but I'd still like you to see them.

dmitrizagidulin commented:
> I disagree that testing documentFetcher() is "out of scope" for this PR. If fetching the resource fails, then the delegation verification doesn't actually accomplish anything at all.

I think you're losing sight of the big picture of this PR. This is a fairly minor fix for a *crashing bug* to solid-server. I did not create `documentFetcher()`. And I do not plan to test it in its current form, in this PR, since it (and several of the functions inside it) are going to be refactored shortly, in a separate branch.

dan-f commented:
> I think you're losing sight of the big picture of this PR. This is a fairly minor fix for a crashing bug to solid-server. I did not create documentFetcher(). And I do not plan to test it in its current form, in this PR, since it (and several of the functions inside it) are going to be refactored shortly, in a separate branch.

Again, I disagree. A patch is not a "fix" unless the patched system is fully tested for the the thing which it claims to "fix".

As one of the maintainers of this Solid implementation, you're free to merge whatever code you want. I'm not going to give the `:+1:` to an implementation that isn't fully tested. I think it's worth pointing out that I spent much of my last week fixing bugs in various solid libraries that did not have test coverage.

dmitrizagidulin commented:
> I think it's worth pointing out that I spent much of my last week fixing bugs in various solid libraries that did not have test coverage.

You say that as if this is some sort of exceptional circumstance, instead of our daily reality as developers. This is a rapidly moving, pre-alpha stage, experimental project. Most the the Solid libraries we use don't have test coverage. And while that should be addressed, as much as possible, as we go along, requiring a full stop until everything gets tested is not reasonable.

dmitrizagidulin commented:
I will refactor `documentFetcher()`, however, and add tests in this PR.

dmitrizagidulin commented:
After further investigation, refactoring and testing `documentFetcher()` requires nontrivial changes to the rest of the server. While these are being addressed, I pushed PR #487 to temporarily address the crashing issue.

RubenVerborgh commented:
We have a different implementation with #646.

melvincarvalho commented:
Please dont delete the dz_fix_delegation branch, there may be useful code there.
